From 70187ebebc62d7ea8cd8b395a763d218a10b9288 Mon Sep 17 00:00:00 2001
From: Xu Qiang <xuqiang36@huawei.com>
Date: Wed, 15 Mar 2023 02:45:07 +0000
Subject: [PATCH] Before entering the second core the following oops occurred
 on an arm64 system.

Mem abort info:
  ESR = 0x96000007
  EC = 0x25: DABT (current EL), IL = 32 bits
  SET = 0, FnV = 0
  EA = 0, S1PTW = 0
Data abort info:
  ISV = 0, ISS = 0x00000007
  CM = 0, WnR = 0
swapper pgtable: 4k pages, 48-bit VAs, pgdp=000000002a1fb000
[ffff800011339280] pgd=000000213ffff003, p4d=000000213ffff003, pud=000000213fffe003, pmd=000000213fffc003, pte=0000000000000000
Internal error: Oops: 96000007 [#1] SMP
CPU: 1 PID: 922 Comm: sh Kdump: loaded Tainted: P           O      5.10.0+ #1
Hardware name: Hisilicon PhosphorHi1910B evb (DT)
pstate: a0400089 (NzCv daIf +PAN -UAO -TCO BTYPE=--)
pc : lpi_update_config+0x194/0x248
lr : lpi_update_config+0x144/0x248
sp : ffff80001ce7b900
x29: ffff80001ce7b900 x28: ffff0020e587a280
x27: 0000000000000000 x26: ffff800011b02000
x25: 0000000000000000 x24: 00000000000021e0
x23: 0000000000000004 x22: ffff800011699000
x21: ffff800011d98000 x20: ffff800011159000
x19: ffff002097bbca80 x18: ffffffffffffffff
x17: 0000000000000000 x16: 0000000000000000
x15: ffff800011159ad8 x14: 0000000060000089
x13: ffff800010183e94 x12: ffff80001ce7ba20
x11: 0000000000000040 x10: ffff800011a49588
x9 : ffff800011a49580 x8 : ffff0020804a0270
x7 : 0000000000000000 x6 : ffff80001169fc40
x5 : 00000000000000a3 x4 : 00000000000001e0
x3 : 0000000000000001 x2 : 0000000000000000
x1 : 0000000000000000 x0 : ffff800011339280
Call trace:
 lpi_update_config+0x194/0x248
 its_mask_irq+0x44/0x50
 irq_chip_mask_parent+0x2c/0x38
 irq_chip_mask_parent+0x2c/0x38
 machine_kexec_mask_interrupts+0x88/0xd8
 machine_crash_shutdown+0x58/0x78
 __crash_kexec+0x78/0x130
 panic+0x208/0x414

After enable CONFIG_ASCEND_INIT_ALL_GICR feature, GICR of
other cores is accessed, which is not supported.

Signed-off-by: Xu Qiang <xuqiang36@huawei.com>
---
 drivers/irqchip/irq-gic-v3-its.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/irqchip/irq-gic-v3-its.c b/drivers/irqchip/irq-gic-v3-its.c
index 6537cedb9..dc38e23ca 100644
--- a/drivers/irqchip/irq-gic-v3-its.c
+++ b/drivers/irqchip/irq-gic-v3-its.c
@@ -1479,7 +1479,7 @@ static void lpi_update_config(struct irq_data *d, u8 clr, u8 set)
 	struct its_device *its_dev = irq_data_get_irq_chip_data(d);
 
 	lpi_write_config(d, clr, set);
-	if (gic_rdists->has_direct_lpi &&
+	if (gic_rdists->has_direct_lpi && !init_all_gicr &&
 	    (is_v4_1(its_dev->its) || !irqd_is_forwarded_to_vcpu(d)))
 		direct_lpi_inv(d);
 	else if (!irqd_is_forwarded_to_vcpu(d))
-- 
2.17.1

